File: ZoneSentry/Models/Realty.cs
namespace ZoneSentry.Models;

public class Realty
{
    public int Id { get; set; }
    
    public int Number { get; set; }
    public int? PaymentPerMonth { get; set; }
    public int? SellCost { get; set; }
    
    public House House { get; set; }
    public int HouseId { get; set; }
    
    public ApplicationUser? Owner { get; set; }
    public bool OwnedByCompany { get; set; }
    
    public RealtyStatus RealtyStatus { get; set; }
    
    public List<RentAgreement> RentAgreements { get; set; }
    public List<RentRequest> RentRequests { get; set; } 
    public List<PurchaseRequest> PurchaseRequests { get; set; } 

    public RentAgreement? CurrentRentAgreement => RentAgreements != null ? RentAgreements.FirstOrDefault(a => a.Date > DateTime.Now && a.ExpirationDate < DateTime.Now) : null;
}

public enum RealtyStatus {
    NotForSale,
    ForRent,
    ForSale,
}

public class RealtyDTO
{
    public int Id { get; set; }
    
    public int Number { get; set; }
    public int? PaymentPerMonth { get; set; }
    public int? SellCost { get; set; }
    
    public int HouseId { get; set; }
    
    public ApplicationUserDTO? Owner { get; set; }
    public bool OwnedByCompany { get; set; }
    
    public List<int> RentAgreements { get; set; }

    public int? CurrentRentAgreement { get; set; }
}

public class RealtyCreate
{
    public int Number { get; set; }
}

public class RealtyUserView
{
    public int Id { get; set; }
    
    public int Number { get; set; }
    public int? PaymentPerMonth { get; set; }
    public int? SellCost { get; set; }
    
    public int HouseId { get; set; }
    
    public ApplicationUserDTO? Owner { get; set; }
    public bool OwnedByCompany { get; set; }
    
    public HouseUserView House { get; set; }
}

File: ZoneSentry/ClientApp/src/api/models/Realty.ts
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

import type { ApplicationUser } from './ApplicationUser';
import type { House } from './House';
import type { PurchaseRequest } from './PurchaseRequest';
import type { RealtyServiceOrder } from './RealtyServiceOrder';
import type { RealtyStatus } from './RealtyStatus';
import type { RentAgreement } from './RentAgreement';
import type { RentRequest } from './RentRequest';

export type Realty = {
    id?: number;
    number?: number;
    paymentPerMonth?: number | null;
    sellCost?: number | null;
    area?: number;
    planUrl?: string | null;
    house?: House;
    houseId?: number;
    owner?: ApplicationUser;
    ownedByCompany?: boolean;
    realtyStatus?: RealtyStatus;
    rentAgreements?: Array<RentAgreement> | null;
    rentRequests?: Array<RentRequest> | null;
    purchaseRequests?: Array<PurchaseRequest> | null;
    services?: Array<RealtyServiceOrder> | null;
    currentRentAgreement?: RentAgreement;
};